Fix an off-by-1 error.

dexterurbane opened the pull request:
Hi Karsten,

There appears to be an off-by-1 error in inner_constraint.cpp. Making this change fixed a segfault I was getting, and thereafter everything seemed normal. But I don't understand the code enough to know if, for example, this change breaks anything.

Annotations below:
```
hess_pos_map[p] = hess_pos_map.size() - 1; // <=== 1st entry into hess_pos_map is -1.
hess_map.push_back(hess_pos_map[p]); // <=== hess_map elems are unsigned, so -1 -> large +ve int.
```

Commit:
Fix an off-by-1 error.

paul-scott commented:
I think this might have been the original cause for #13 (the segfault part), and it sounds consistent with the issue in #15 (segfault only when exact hessian calcs used).

Commit:
Make previous change more robust on different compilers.

Previous change fixed a problem on MacOS/clang++, but created another
problem on linux/g++. There appears to be some compiler related issue
with the order of evaluation in the statement. Implement more robust
syntax for change.

paul-scott commented:
Bit of clarification, it seems like different compilers would evaluate the map [ ] operator and the RHS in different orders (thought this would be standard).  The latest change should make it unambiguous.

dexterurbane commented:
You'll notice that I updated my fork (latest = [97e037c](https://github.com/stanle/madopt/pull/16/commits/97e037c699510d6164b858bc0fe95ae9689ce7a6)). Basically, some nasty order of evaluation stuff was going on, between assignment, operator[] and size(). Different compilers were behaving differently, so I changed the code to something more robust.

Cheers, Dan
